api/src/main/java/org/openmrs/module/locationbasedaccess/EncounterSearchAdviser.java

```diff
@@ -0,0 +1,137 @@
+
+    private static final Log log = LogFactory.getLog(EncounterSearchAdviser.class);
+
+    @Override
+    public boolean matches(Method method, Class targetClass) {
+        if (method.getName().equals("getEncounter")) {
+            return true;
+        }
+        else if (method.getName().equals("getEncounterByUuid")) {
+            return true;
+        }
+        else if (method.getName().equals("getEncounters")) {
+            return true;
+        }
+        else if (method.getName().equals("getEncountersByPatientId")) {
+            return true;
+        }
+        else if (method.getName().equals("getEncountersByPatient")) {
+            return true;
+        }
+        else if (method.getName().equals("getEncountersByVisit")) {
+            return true;
+        }
+        else if (method.getName().equals("getEncountersNotAssignedToAnyVisit")) {
+            return true;
+        }
+        else if (method.getName().equals("getEncountersByVisitsAndPatient")) {
+            return true;
+        }
+        else if (method.getName().equals("getAllEncounters")) {
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public Advice getAdvice() {
+        return new EncounterSearchAdvise();
+    }
+
+    private class EncounterSearchAdvise implements MethodInterceptor {
+        public Object invoke(MethodInvocation invocation) throws Throwable {
+            if (Context.getAuthenticatedUser() == null) {
+                return null;
+            }
+
+            Object object = invocation.proceed();
+            if (Daemon.isDaemonUser(Context.getAuthenticatedUser()) || Context.getAuthenticatedUser().isSuperUser()) {
+                return object;
+            }
+
+            Integer sessionLocationId = Context.getUserContext().getLocationId();
+            if (sessionLocationId != null) {
+                String sessionLocationUuid = Context.getLocationService().getLocation(sessionLocationId).getUuid();
+
+                if (object instanceof List) {
+                    List<Encounter> encounterList = (List<Encounter>) object;
+                    object = removeEncountersIfNotBelongToGivenLocation(encounterList, sessionLocationUuid);
+                } else if (object instanceof Map) {
+                    Map<Integer, List<Encounter>> encounterMap = (Map<Integer, List<Encounter>>) object;
+                    Iterator<Map.Entry<Integer, List<Encounter>>> mapIterator = encounterMap.entrySet().iterator();
+                    while (mapIterator.hasNext()) {
+                        Map.Entry<Integer, List<Encounter>> entry = mapIterator.next();
+                        List<Encounter> encounterList = entry.getValue();
+                        entry.setValue(removeEncountersIfNotBelongToGivenLocation(encounterList, sessionLocationUuid));
+                        //TODO: remove the entry from the map, if the encounter list is empty and update the map index
```

Comment:
When are you planning to address this TODO?

Reply:
Here If I remove this value from the Map, then I need to update the integer index as well. So It should be done using some preferred way. It's not very much needed for the first use case. So decided to do it later.